lib/prism.js

```diff
@@ -0,0 +1,92 @@
+'use strict';
+
+const Prism = require('prismjs');
+const prismLoadLanguages = require('prismjs/components/');
+
+// https://github.com/PrismJS/prism/issues/2145
+const prismComponents = require('prismjs/components');
+const prismSupportedLanguages = Object.keys(prismComponents.languages);
+
+const escapeHTML = require('./escape_html');
+
+/**
+ * Wrapper of Prism.highlight()
+ * @param {String} code
+ * @param {String} language
+ */
+function prismHighlight(code, language) {
+  // Prism has not load the language pattern
+  if (!Prism.languages[language] && prismSupportedLanguages.includes(language)) prismLoadLanguages(language);
+
+  if (Prism.languages[language]) {
+    // Prism escapes output by default
+    return Prism.highlight(code, Prism.languages[language], language);
+  }
+
+  // Current language is not supported by Prism, return origin code;
+  return escapeHTML(code);
+}
+
+/**
+ * Generate line number required HTML snippet
+ * @param {String} code - Highlighted code
+ */
+function lineNumberUtil(code) {
+  const matched = code.match(/\n(?!$)/g);
+  const num = matched ? matched.length + 1 : 1;
+  const lines = new Array(num + 1).join('<span></span>');
+
+  return `<span aria-hidden="true" class="line-numbers-rows">${lines}</span>`;
+}
+
+function replaceTabs(str, tab) {
+  return str.replace(/^\t+/gm, match => {
+    let result = '';
+
+    for (let i = 0, len = match.length; i < len; i++) {
+      result += tab;
+    }
+
+    return result;
+  });
+}
+
+function PrismUtil(str, options = {}) {
+  if (typeof str !== 'string') throw new TypeError('str must be a string!');
+
+  const {
+    lineNumber = true,
+    lang = 'none',
+    tab,
+    isPreprocess = true
+  } = options;
+
+  // To be consistent with highlight.js
+  const language = lang === 'plain' || lang === 'none' ? 'none' : lang;
+
+  const preTagClassArr = [];
+
+  if (lineNumber) preTagClassArr.push('line-numbers');
+  preTagClassArr.push(`language-${language}`);
+
+  if (tab) str = replaceTabs(str, tab);
+
+  const startTag = `<pre class="${preTagClassArr.join(' ')}"><code class="language-${language}">`;
+  const endTag = '</code></pre>';
+
+  let parsedCode = '';
+
+  if (language === 'none' || !isPreprocess) {
```

Comment:
https://github.com/hexojs/hexo/issues/4010#issuecomment-568906344

> During [hexojs/hexo-util#168](https://github.com/hexojs/hexo-util/pull/168), I have noticed PrismJS is mainly designed for working at browser side. Many features likes "highlight specific lines", "The first line number" only work in browser.
> 
> So if we are going to bring PrismJS to Hexo, IMHO we should address two options:
> * `browser` mode. In this mode Hexo will produce HTML snippet that will could be processed by `prism.js` & `prism.css` in browser. The full features of PrismJS could be used then, and it will be compatible with all options of current `include_code()` and `code()` helper.
> * `preprocess` mode. In this mode Hexo will produce prism-parsed HTML snippet. Then theme/user only need to add `prism.css`. However, many features will no longer works under `preprocess` mode thus will not be compatible with `include_code()` and `code()` helper.